tty_test.go

```diff
@@ -69,110 +69,122 @@ func TestReconnectTTY(t *testing.T) {
 	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
 	defer cancel()
 
-	ws, server := mockConn(ctx, t, &Options{
+	ws1, server1 := mockConn(ctx, t, &Options{
 		ReconnectingProcessTimeout: time.Second,
 	})
-	defer server.Close()
+	defer server1.Close()
 
 	command := Command{
 		ID:      uuid.NewString(),
 		Command: "sh",
 		TTY:     true,
 		Stdin:   true,
 	}
-	execer := RemoteExecer(ws)
-	process, err := execer.Start(ctx, command)
+	execer1 := RemoteExecer(ws1)
+	process1, err := execer1.Start(ctx, command)
 	assert.Success(t, "start sh", err)
 
 	// Write some unique output.
 	echoCmd := "echo test:$((1+1))"
 	data := []byte(echoCmd + "\r\n")
-	_, err = process.Stdin().Write(data)
+	_, err = process1.Stdin().Write(data)
 	assert.Success(t, "write to stdin", err)
 	expected := []string{echoCmd, "test:2"}
 
-	findEcho := func(expected []string) bool {
-		scanner := bufio.NewScanner(process.Stdout())
-	outer:
-		for _, str := range expected {
-			for scanner.Scan() {
-				line := scanner.Text()
-				t.Logf("bash tty stdout = %s", line)
-				if strings.Contains(line, str) {
-					continue outer
-				}
-			}
-			return false // Reached the end of output without finding str.
-		}
-		return true
-	}
-
-	assert.True(t, "find echo", findEcho(expected))
+	assert.True(t, "find echo", findEcho(t, process1, expected))
 
 	// Test disconnecting then reconnecting.
-	ws.Close(websocket.StatusNormalClosure, "disconnected")
-	server.Close()
+	process1.Close()
```

Comment:
Does this change the semantics a little?  Before it seemed more like if the socket closed unexpectedly due to external circumstances while now it feels more like a proper close but I am not sure if this distinction matters or if that was even the original intent.

Reply:
I don't think the original represented an unexpected close, as it was all client side.  If we wanted to test for an unexpected close, we'd need to do it server-side, or possibly kill the net.Conn the websocket is running over.

But, the lesson of this whole PR is that it's not threadsafe to call `websocket.Close()` while we are listening for a remote process. I feel like this gets us close to the original intent of the test, which is to show that the server is maintaining the process even if the client disconnects.

Reply:
That seems fair, I think where I was driving at is whether we only need the `server.Close()` depending on whether the original intent was to include an unexpected disconnect (on account of being threadsafe) but it makes sense to me that what we really care about is that the process is reconnecting.  :+1: